### PR TITLE
Issue 24132 - ImportC: Add support for wide-chars.

### PR DESCRIPTION
**Pull request by LightBender**
Fix Issue 24132. Add wchar/dchar to C Keywords list. Use #defines to convert C wide-chars to wchar/dchar

### CODE REVIEW

File: druntime/src/importc.h [hunk 1]
```diff
@@ -164,3 +164,8 @@
 #if __APPLE__
 #undef __SIZEOF_INT128__
 #endif
+
+// Wide-char compatibility
+#define wchar_t wchar
```
Comment (thewilsonator) [inline druntime/src/importc.h:169]: this is wrong, see https://github.com/dlang/dmd/blob/master/druntime/src/core/stdc/stddef.d#L35-L40

File: compiler/src/dmd/tokens.d [hunk 2]
```diff
@@ -607,7 +607,7 @@ static immutable TOK[TOK.max + 1] Ckeywords =
     with (TOK)
     {
         TOK[TOK.max + 1] tab = identifier;  // default to identifier
-        enum Ckwds = [ auto_, break_, case_, char_, const_, continue_, default_, do_, float64, else_,
+        enum Ckwds = [ auto_, break_, case_, char_, wchar_, dchar_, const_, continue_, default_, do_, float64, else_,
```
Comment (thewilsonator) [inline compiler/src/dmd/tokens.d:610]: I suspect the correct way to do this is add `wchar_t`/`char16_t`/`char32_t` as tokens and translate them in the parser, not hack the lexer like this.

**Review by thewilsonator — CHANGES_REQUESTED** [review]

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Let's not add D keywords to C.

File: compiler/src/dmd/tokens.d (hunk 2 above)
Comment (LightBender) [reply compiler/src/dmd/tokens.d:610]: I considered that, however, these #defines in C. Something like:` #define wchar_t unsigned short` so they won't even enter the D compiler as those types.

**Comment by LightBender** [comment]
> Let's not add D keywords to C.

Have another way to signify that we want a `wchar` instead of an `unsigned short`?

**Comment by thewilsonator** [comment]
> Have another way to signify that we want a wchar instead of an unsigned short?

Leaving aside the fact that posix wants 32-bit wchar not 16 bit for a moment. You want to add `wchar_t/char16_t/char32_t` tokens and then in the parser translate them to `wchar/dchar` as appropriate.

**Comment by ibuclaw** [comment]
According to C11 (and just checked C2X too) `7.19 Comon definitions`.

> 1. The header **<stddef.h>** defines the following macros and declares the following types. Some are
also defined in other headers, as noted in their respective subclauses.
> 2. The types are
> ...
> `wchar_t`
> which is an integer type whose range of values can represent distinct codes for all members of the
largest extended character set specified among the supported locales; the null character shall have
the code value zero. Each member of the basic character set shall have a code value equal to its
value when used as the lone character in an integer character constant if an implementation does
not define __STDC_MB_MIGHT_NEQ_WC__ .

This is a library type, not compiler.

**Comment by LightBender** [comment]
> This is a library type, not compiler.

That is correct. However, when using ImportC headers, we don't want to work with integer types, they are the same at the ABI level, but that is incorrect semantically from the C API's that use wide-chars, which is most of the Windows API (as an example). I'm trying to preserve the implied type information so that the tool-chain can correctly utilize the expected types.

**Comment by ibuclaw** [comment]
Looks like the preprocessor should ensure it gets expanded appropriately - from the `ginclude/stddef.h` header I have for reference:
```
#ifndef __WCHAR_TYPE__  // The preprocessor predefines this as necessary per target
#define __WCHAR_TYPE__ int
#endif
typedef __WCHAR_TYPE__ wchar_t;
```

**Comment by ibuclaw** [comment]
> but that is incorrect semantically from the C API's that use wide-chars

C doesn't apply any special semantics to wchar/charN types. So what does this mean?

**Comment by LightBender** [comment]
> Looks like the preprocessor should ensure it gets expanded appropriately - from the `ginclude/stddef.h` header I have for reference:

Kinda, `__WCHAR_TYPE__` is not defined in C. As far as Walter and I have found, it's only defined when compiling as C++. So we'll have to normalize all of these with the pre-processor. See updated PR.

**Comment by LightBender** [comment]
> > but that is incorrect semantically from the C API's that use wide-chars
> 
> C doesn't apply any special semantics to wchar/charN types. So what does this mean?

It's a case where D has more capability than C, so with C they just use the pre-processor, whereas with D it's a native type. By up-converting to a native type we provided the downstream tooling, and the devs with more information about what the method parameter actually requires.

**Comment by ibuclaw** [comment]
> > Looks like the preprocessor should ensure it gets expanded appropriately - from the `ginclude/stddef.h` header I have for reference:
> 
> Kinda, `__WCHAR_TYPE__` is not defined in C. As far as Walter and I have found, it's only defined when compiling as C++. So we'll have to normalize all of these with the pre-processor. See updated PR.

It's a common predefined macro on at least one vendor compiler (https://godbolt.org/z/cG55dYK5n)

File: compiler/src/dmd/tokens.d [hunk 3]
```diff
@@ -268,6 +268,8 @@ enum TOK : ubyte
     _Noreturn,
     _Static_assert,
     _Thread_local,
+	wchar_t,
+    char32_t,
```
Comment (thewilsonator) [inline compiler/src/dmd/tokens.d:272]: need to add these to `extern (D) private static immutable string[TOK.max + 1] tochars =`

**Comment by thewilsonator** [comment]
> It's a common predefined macro on at least one vendor compiler (https://godbolt.org/z/cG55dYK5n)

clang defines it too, except clang seems to have it be unsigned. Also tested works: icc, icx. Seems to be only mdvc that doesn't do that.

**Comment by ibuclaw** [comment]
> > > but that is incorrect semantically from the C API's that use wide-chars
> > 
> > 
> > C doesn't apply any special semantics to wchar/charN types. So what does this mean?
> 
> It's a case where D has more capability than C, so with C they just use the pre-processor, whereas with D it's a native type. By up-converting to a native type we provided the downstream tooling, and the devs with more information about what the method parameter actually requires.

Yeah, no. I don't think you should be applying D semantics to C types.  D's `wchar` and `dchar` are both _fixed_ and _unsigned_. C's character types could be any size (16 or 32, typically whatever is the natural word size) or signness - a quick dedup'd grep finds the following across ~100 or so ABIs.
```
"int"
"long int"
"long unsigned int"
"short int"
"short unsigned int"
"unsigned int"
```

**Comment by LightBender** [comment]
> Yeah, no. I don't think you should be applying D semantics to C types. D's `wchar` and `dchar` are both _fixed_ and _unsigned_. C's character types could be any size (16 or 32, typically whatever is the natural word size) or signness - a quick dedup'd grep finds the following across ~100 or so ABIs.
> 
> ```
> "int"
> "long int"
> "long unsigned int"
> "short int"
> "short unsigned int"
> "unsigned int"
> ```
> 
> And I don't even know what size any of those type really are.

Does not D already map wchar/dchar those ABIs correctly when using extern(C)? I recall being told that it does. So I fail to see the problem here. We gain information for no additional cost.

**Comment by ibuclaw** [comment]
> Does not D already map wchar/dchar those ABIs correctly when using extern(C)? I recall being told that it does. If so, I fail to see the problem here. We gain information for no additional cost.

There's nothing to map to in extern(C). Types aren't encoded into the symbol, and you could be forgiven for getting the size wrong because most ABIs do integer promotion anyway (everything is an `int`) - except for when it comes to layout in a struct.

**Comment by ibuclaw** [comment]
> this is wrong, see https://github.com/dlang/dmd/blob/master/druntime/src/core/stdc/stddef.d#L35-L40

Druntime is also wrong. But _sic vita est_.

**Comment by thewilsonator** [comment]
> Druntime is also wrong. But sic vita est.

What is it supposed to be then?

**Comment by ibuclaw** [comment]
> > Druntime is also wrong. But sic vita est.
> 
> What is it supposed to be then?

Broken up per CPU.

**Comment by LightBender** [comment]
I'm going to close this and attack the problem a different way.

**Pull request closed without merging**